src/state.py

```diff
@@ -5,11 +5,54 @@ def __init__(self):
     def add_student(self, student):
         self.students.append(student)
 
+    #Given two students and a subject, trades their classes for that given subject
+    def trade_classes(self,student1,student2,subject_name):
+        student1_class = student1.get_class_for_subject(subject_name)
+        student2_class = student2.get_class_for_subject(subject_name)
+        student1.set_class_for_subject(subject_name, student2_class)
+        student2.set_class_for_subject(subject_name, student1_class)
+        #student1.remove_subject_target(subject_name) #E suposto ele remover os targets a medida que vai trocando? Ou e usado para a funçao de fitness?
+        #student2.remove_subject_target(subject_name)
+
+    def generate_neighbour(self):
+        for student_index,student in enumerate(self.students,start=1): # select a student
+
+
+            for subject_name,target_classes in student.subject_targets.items(): # select a subject name and its respective  target classes
+
+                student_class = student.get_class_for_subject(subject_name) # get the current class of the first student for that subject
+
+
+                for target_class in target_classes: # select, for that subject, a possible target class
+
+                    for trader_student in self.students[student_index:]: # find a trader student
+                        if trader_student == student:
+                            continue
+                        
+                        trader_class = trader_student.get_class_for_subject(subject_name) # get the class of the trader student, for that subject
+
+                        if trader_class is None or trader_class != target_class: # if target student does not have the desired class for that subject, continue
+                            continue
+                  
+                        trader_targets = trader_student.get_targets_for_subject(subject_name) # get the targets for that subject, for the trader student
+                        
+                        if trader_targets is not None and student_class in trader_targets:
+                            self.trade_classes(student,trader_student,subject_name) #trades the students classes for that subject
+                            break
+                                     
+                        #Afinal os give in so vao ser feitos depois de já se ter verificado os targets todos
+                        trader_giveins = trader_student.get_giveins_for_subject(subject_name) # get the giveins for that subject, for the trader student
+                        #if trader_giveins is not None and student_class in trader_giveins:
+                            # TODO: GENERATE NEW STATE, SWITCH CLASSES AND FINISH
```

Comment:
This function should actually create new states, instead of changing the current state. 

Instead of breaking or returning, trying making this a generator function, using yield on every newly generated object.
  
This will allow for the main cycle to evaluate each neighbor as it is generated, instead of waiting for all of them to be generated before evaluation.

Reply:
**If you find yourselves asking why we should use a generator function**
Since we are implementing a simple Hill Climbing, the current state should change to the first neighbor who has a better fitness.
This means that if we evaluate a neighbor as it is generated and find that he is better than the current state, we needn't generate the next neighbor. 
Therefore, a generator function will allow to only generate the needed amount of states.

Reply:
Already made the changes requested